Adding documentation for F5 NGINX Ingress controller

jasonwilliams14 opened the pull request:
This is a guide on how to use F5 NGINX Ingress controller with open service mesh. The document works with both NGINX Ingress Controller OSS and NGINX Ingress Controller Plus.
The document also covers how to setup this integration using the sidecar method and the ingressBackend method.

jasonwilliams14 committed:
Adding documentation for F5 NGINX Ingress controller

Signed-off-by: Jason Williams <jason.williams14@gmail.com>

phillipgibson commented:
Thanks @jasonwilliams14 for submitting this. I'll take a look at this for the upcoming week.

jasonwilliams14 commented:
> Thanks @jasonwilliams14 for submitting this. I'll take a look at this for the upcoming week.

You bet. I might add or update some of the doc. There are a few other tests that I am curious to test. Will update if anything comes up.

phillipgibson commented:
@jasonwilliams14 can you update to how nginx is installed? The doc jumps right over that.

jasonwilliams14 commented:
@phillipgibson I updated the doc to cover the installation of F5 NGINX Ingress controller.
Let me know if you need anything else.

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+
+### Mark F5 NGINX Ingress controller namespace for sidecar injection
+
+*NOTE:* Depending on how you install NGINX Ingress controller, you might need to create the `namespae`. For example, if you are using manifests to install NGINX Ingress controller, you can complete all of the steps on our documentation page, *EXCEPT*, actually deploying NGINX Ingress controller. This is because, when using the sidecar approach, OSM needs to "manage" the namespace so it knows what ...
```
```suggestion
*NOTE:* Depending on how you install NGINX Ingress Controller, you might need to create the `namespace`. For example, if you are using manifests to install NGINX Ingress Controller, you can complete all of the steps on our documentation page, *EXCEPT*, actually deploying NGINX Ingress Controller. This is because, when using the sidecar approach, OSM needs to "manage" the namespace so it knows what `namespaces` it needs to inject sidecars into.
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+2/2 shows we have two containers in the NGINX Ingress controller pod: NGINX Ingress and Envoy
+
+
+Configure your NGINX VirtualServer yaml to similar below
```
```suggestion
Configure your NGINX VirtualServer yaml similar to below
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+
+### mTLS Setup
+
+To enable mTLS for NGINX Ingress Controller and OSM, you need to configure the `IngressBackend` API to use `https` as the backend protocol as and trigger OSM to issue a certificate. NGINX will use this certificate to proxy HTTPS connections to the TLS backends. The client certificate and certificate authority (CA) certificate will be stored in a Kubernetes secret that NGINX will use for authentic ...
```
```suggestion
To enable mTLS for NGINX Ingress Controller and OSM, you need to configure the `IngressBackend` API to use `https` as the backend protocol and trigger OSM to issue a certificate. NGINX will use this certificate to proxy HTTPS connections to the TLS backends. The client certificate and certificate authority (CA) certificate will be stored in a Kubernetes secret that NGINX will use for authentication."
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+
+Edit `osm-ca-bundle-secret.yaml`
+Remove the `private.key` section under `data.`
+Change the `namespace` field to your nginx-ingress location
```
```suggestion
Change the `namespace` field to your nginx-ingress namespace
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+
+Edit `osm-nginx-client-cert.yaml`
+Remove the `ca.crt` in the `data` section
+Change the namespace to the nginx-ingress namespace.
```
```suggestion
Change the `namespace` to the nginx-ingress namespace.
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
````diff
@@ -0,0 +1,420 @@
+kubectl apply -f osm-nginx-client-cert.yaml
+
+```
+Ensure the secrets exisit in the `nginx-ingress` namespace:
````
```suggestion
Ensure the secrets exist in the `nginx-ingress` namespace:
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
````diff
@@ -0,0 +1,420 @@
+osm-ca-bundle           nginx.org/ca        1      23m
+```
+
+We now need to create our CRDs (virtualServer and policy).
````
```suggestion
We now need to create our CRDs (VirtualServer and Policy).
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+
+We now need to create our CRDs (virtualServer and policy).
+Here is the `policy` resource that holds the mTLS information.
+Make sure you apply the `policy` or the  mTLS connection will not work. (required for virtualServer)
```
```suggestion
Make sure you apply the `policy` or the mTLS connection will not work (required for VirtualServer).
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
````diff
@@ -0,0 +1,420 @@
+    sslName: httpbin.httpbin.cluster.local
+```
+
+Here is an example `virtualServer` resource as well as the `ingressBackend`.
````
```suggestion
Here is an example `VirtualServer` resource as well as the `IngressBackend`.
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
````diff
@@ -0,0 +1,420 @@
+    name: nginx-ingress.nginx-ingress.cluster.local
+```
+
+Once these are applied, verify they are valid (virtualServer) and committed (ingressBackend):
````
```suggestion
Once these are applied, verify they are valid (VirtualServer) and committed (IngressBackend):
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+## This document outlines how to integrate F5 NGINX Ingress Controller with Open Service Mesh (OSM)
+
+Open Service Mesh will work with both versions of [F5 NGINX Ingress controller](https://github.com/nginxinc/kubernetes-ingress): the free as well as the NGINX Plus versions.
```
```suggestion
Open Service Mesh will work with both versions of [F5 NGINX Ingress Controller](https://github.com/nginxinc/kubernetes-ingress): the free as well as the NGINX Plus versions.
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+
+Open Service Mesh will work with both versions of [F5 NGINX Ingress controller](https://github.com/nginxinc/kubernetes-ingress): the free as well as the NGINX Plus versions.
+
+Below is a link to the official F5 NGINX Ingress controller documentation.    
```
```suggestion
Below is a link to the official F5 NGINX Ingress Controller documentation.
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+Open Service Mesh will work with both versions of [F5 NGINX Ingress controller](https://github.com/nginxinc/kubernetes-ingress): the free as well as the NGINX Plus versions.
+
+Below is a link to the official F5 NGINX Ingress controller documentation.    
+[F5 NGINX Ingress controller](https://docs.nginx.com/nginx-ingress-controller/)
```
```suggestion
[F5 NGINX Ingress Controller](https://docs.nginx.com/nginx-ingress-controller/)
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+1. Injecting an envoy sidecar directly with NGINX Ingress Controller.
+2. Using the Open Service Mesh `ingressBackend` "proxy" feature.
+
+# First step is to install F5 NGINX Ingress controller. Here the links to the install guides:
```
```suggestion
# First step is to install F5 NGINX Ingress Controller. Here are the links to the install guides:
```

jasonwilliams14 replied:
@phillipgibson Definitely. I can add that to the doc in the install portion of the document. Does proving a small snippet of the annotation in the `values.yaml` work? That would give users an idea of what adding that annotation to the helm deployment would look like.

phillipgibson replied:
That works. I was trying to repro with this:
```
helm install nginx-osm-demo nginx-stable/nginx-ingress --namespace nginx-osm-ing --set-string controller.podAnnotations.'openservicemesh.io/inbound-port-exclusion-list={80, 443}'
```

But I don't think it created the annotations needed.

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
`````diff
@@ -0,0 +1,420 @@
+````
+
+
+# NGINX Ingress controller and OSM with sidecar injected 
`````
```suggestion
# NGINX Ingress Controller and OSM with sidecar injected 
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
````diff
@@ -0,0 +1,420 @@
+osm install --mesh-name osm-nginx --osm-namespace osm-system
+```
+
+### Mark F5 NGINX Ingress controller namespace for sidecar injection
````
```suggestion
### Mark F5 NGINX Ingress Controller namespace for sidecar injection
```

phillipgibson replied:
With my earlier comment on showing a namespace for the install, it makes this step easier to follow to know which namespace OSM is to manage.

jasonwilliams14 replied:
@phillipgibson I will make sure to call that out with the install portion, highlighting the namespace that F5 NGINX Ingress will be installed into.

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+
+*NOTE:* Depending on how you install NGINX Ingress controller, you might need to create the `namespae`. For example, if you are using manifests to install NGINX Ingress controller, you can complete all of the steps on our documentation page, *EXCEPT*, actually deploying NGINX Ingress controller. This is because, when using the sidecar approach, OSM needs to "manage" the namespace so it knows what ...
+
+Next thing we need to do is install OSM into the `NGINX Ingress controller` namespace so that the `envoy` sidecar will be injected into NGINX Ingress controller.
```
```suggestion
The next thing we need to do is install OSM into the NGINX Ingress Controller namespace so that the Envoy sidecar will be injected into NGINX Ingress Controller.
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
````diff
@@ -0,0 +1,420 @@
+        openservicemesh.io/inbound-port-exclusion-list: "80,443"
+```
+
+This annotation is *required* when injecting envoy sidecar into NGINX Ingress controller.
````
```suggestion
This annotation is *required* when injecting the Envoy sidecar into NGINX Ingress Controller.
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
````diff
@@ -0,0 +1,420 @@
+nginx-ingress-7b9557ddc6-zw7l5   2/2     Running   1 (5m8s ago)   5m19s
+```
+
+2/2 shows we have two containers in the NGINX Ingress controller pod: NGINX Ingress and Envoy
````
```suggestion
2/2 shows we have two containers in the NGINX Ingress Controller pod: NGINX Ingress and Envoy
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+[Using Manifests to install NGINX Ingress](https://docs.nginx.com/nginx-ingress-controller/installation/installation-with-manifests/)
+
+
+When using the sidecar method, ensure that you add the correct annotations listed below. This ensures proper integration of NGINX Ingress Controller with the envoy sidecar proxy.
```
```suggestion
When using the sidecar method, ensure that you add the correct annotations listed below. This ensures proper integration of NGINX Ingress Controller with the Envoy sidecar proxy.
```

aknot242 commented on content/en/docs/demos/f5_nginx_ingress_controller.md:
```diff
@@ -0,0 +1,420 @@
+`InboundPortExclusionList` defines a global list of ports to exclude from inbound traffic interception by the sidecar proxy.
+
+
+### Verify that the envoy sidecar has been *injected* into NGINX Ingress
```
```suggestion
### Verify that the Envoy sidecar has been *injected* into NGINX Ingress
```

phillipgibson commented:
> @phillipgibson I updated the doc to cover the installation of F5 NGINX Ingress controller. Let me know if you need anything else.

Taking a look today.

jasonwilliams14 commented:
> > @phillipgibson I updated the doc to cover the installation of F5 NGINX Ingress controller. Let me know if you need anything else.
> 
> Taking a look today.

@phillipgibson I updated the PR today and added more detail about how to install F5 NGINX Ingress controller.

phillipgibson committed:
Update f5_nginx_ingress_controller.md

I added the necessary document front matter and was trying to properly format each section title. The section titles will need to be configured correctly to ensure the content is associated correctly with the proper formatting.

phillipgibson commented:
@jasonwilliams14  please see this preview to see if the page sections are formatted correctly.  https://deploy-preview-451--osm-docs.netlify.app/docs/demos/f5_nginx_ingress_controller/

jasonwilliams14 commented:
> @jasonwilliams14 please see this preview to see if the page sections are formatted correctly. https://deploy-preview-451--osm-docs.netlify.app/docs/demos/f5_nginx_ingress_controller/

@phillipgibson This looks good. Thank you very much!

phillipgibson commented:
@jasonwilliams14 I'm going to run through this for verification. You'll need to sign your commits with DCO. Have a look at https://github.com/openservicemesh/osm-docs/blob/main/CONTRIBUTING.md#sign-your-work for the procedure.

jasonwilliams14 commented:
> @jasonwilliams14 I'm going to run through this for verification. You'll need to sign your commits with DCO. Have a look at https://github.com/openservicemesh/osm-docs/blob/main/CONTRIBUTING.md#sign-your-work for the procedure.

I believe my commit is signed as I reviewed the commit history. It does not look like your commit is signed?

I can squash the commits and sign it if that is helpful. Happy to do so. Thank you.
